nano/node/transport/tcp_server.cpp

```diff
@@ -518,47 +535,71 @@ void nano::transport::tcp_server::handshake_message_visitor::node_id_handshake (
 	{
 		return;
 	}
+
 	if (node->flags.disable_tcp_realtime)
 	{
-		node->logger.debug (nano::log::type::tcp_server, "Handshake attempted with disabled realtime TCP ({})", nano::util::to_str (server->remote_endpoint));
+		node->stats.inc (nano::stat::type::tcp_server, nano::stat::detail::handshake_error);
+		node->logger.debug (nano::log::type::tcp_server, "Handshake attempted with disabled realtime TCP ({})", fmt::streamed (server->remote_endpoint));
 
-		// Stop invalid handshake
-		server->stop ();
-		return;
+		result = status::abort;
+		return; // Abort
 	}
+	if (!message.query && !message.response)
+	{
+		node->stats.inc (nano::stat::type::tcp_server, nano::stat::detail::handshake_error);
+		node->logger.debug (nano::log::type::tcp_server, "Invalid handshake message received ({})", fmt::streamed (server->remote_endpoint));
 
-	if (message.query && server->handshake_query_received)
+		result = status::abort;
+		return; // Abort
+	}
+	if (message.query && server->handshake_received) // Second handshake message should be a response only
 	{
-		node->logger.debug (nano::log::type::tcp_server, "Detected multiple handshake queries ({})", nano::util::to_str (server->remote_endpoint));
+		node->stats.inc (nano::stat::type::tcp_server, nano::stat::detail::handshake_error);
+		node->logger.debug (nano::log::type::tcp_server, "Detected multiple handshake queries ({})", fmt::streamed (server->remote_endpoint));
 
-		// Stop invalid handshake
-		server->stop ();
-		return;
+		result = status::abort;
+		return; // Abort
 	}
 
-	server->handshake_query_received = true;
+	node->stats.inc (nano::stat::type::tcp_server, nano::stat::detail::node_id_handshake, nano::stat::dir::in);
 
-	node->logger.debug (nano::log::type::tcp_server, "Handshake query received ({})", nano::util::to_str (server->remote_endpoint));
+	server->handshake_received = true;
+
+	node->logger.debug (nano::log::type::tcp_server, "Handshake query received ({})", fmt::streamed (server->remote_endpoint));
```

Comment:
Should this be within if (message.query) since it says it's a query? Or reworded to handshake message?